pkg/controller/storage/capability/capability_controller.go

```diff
@@ -0,0 +1,317 @@
+			c.workQueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workQueue but got %#v", obj))
+			return nil
+		}
+		if err := c.syncHandler(key); err != nil {
+			c.workQueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		c.workQueue.Forget(obj)
+		klog.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+// When creating a new storage class, the controller will create a new storage capability object.
+// When updating storage class, the controller will update or create the storage capability object.
+// When deleting storage class, the controller will delete storage capability object.
+func (c *StorageCapabilityController) syncHandler(key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get StorageClass
+	storageClass, err := c.storageClassLister.Get(name)
+	klog.V(4).Infof("Get storageClass %s: entity %v", name, storageClass)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_, err = c.storageClassCapabilityLister.Get(name)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					return nil
+				}
+				return err
+			}
+			return c.storageClassCapabilityClient.StorageV1alpha1().StorageClassCapabilities().Delete(name, &metav1.DeleteOptions{})
+		}
+		return err
+	}
+
+	// Get SnapshotClass
+	snapshotClassCreated := true
+	_, err = c.snapshotClassLister.Get(storageClass.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			snapshotClassCreated = false
+		} else {
+			return err
+		}
+	}
+
+	// Get exist StorageClassCapability
+	storageClassCapabilityExist, err := c.storageClassCapabilityLister.Get(storageClass.Name)
+	if errors.IsNotFound(err) {
+		// If the resource doesn't exist, we'll create it
+		klog.V(4).Infof("Create StorageClassProvisioner %s", storageClass.GetName())
+		storageClassCapabilityCreate := &crdapi.StorageClassCapability{ObjectMeta: metav1.ObjectMeta{Name: storageClass.Name}}
+		err = c.addSpec(&storageClassCapabilityCreate.Spec, storageClass, snapshotClassCreated)
+		if err != nil {
+			return err
+		}
+		klog.V(4).Info("Create StorageClassCapability: ", storageClassCapabilityCreate)
+		_, err = c.storageClassCapabilityClient.StorageV1alpha1().StorageClassCapabilities().Create(storageClassCapabilityCreate)
+		return err
+	}
+	if err != nil {
+		return err
+	}
+
+	// If the resource exist, we can update it.
+	storageClassCapabilityUpdate := storageClassCapabilityExist.DeepCopy()
+	err = c.addSpec(&storageClassCapabilityUpdate.Spec, storageClass, snapshotClassCreated)
+	if err != nil {
+		return err
+	}
+	klog.V(4).Info("Update StorageClassCapability: ", storageClassCapabilityUpdate)
+	_, err = c.storageClassCapabilityClient.StorageV1alpha1().StorageClassCapabilities().Update(storageClassCapabilityUpdate)
```

Comment:
better to do a reflect.compare before updating, if they are identical, an update will change the object resource version which will lead to another object updated event

Reply:
OK, I will fix it.